segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -202,30 +204,47 @@ public void close() {
      */
     @Override
     public void initialize(Duration timeout) throws DurableDataLogException {
+        List<Long> ledgersToDelete;
         synchronized (this.lock) {
             Preconditions.checkState(this.writeLedger == null, "BookKeeperLog is already initialized.");
             assert this.logMetadata == null : "writeLedger == null but logMetadata != null";
 
             // Get metadata about the current state of the log, if any.
-            LogMetadata metadata = loadMetadata();
+            LogMetadata oldMetadata = loadMetadata();
 
-            // Fence out ledgers.
-            if (metadata != null) {
-                Ledgers.fenceOut(metadata.getLedgers(), this.bookKeeper, this.config, this.traceObjectId);
+            if (oldMetadata != null) {
+                // Fence out ledgers.
+                val emptyLedgerIds = Ledgers.fenceOut(oldMetadata.getLedgers(), this.bookKeeper, this.config, this.traceObjectId);
+
+                // Update Metadata to reflect those newly found empty ledgers.
+                oldMetadata = oldMetadata.updateLedgerStatus(emptyLedgerIds);
             }
 
             // Create new ledger.
             LedgerHandle newLedger = Ledgers.create(this.bookKeeper, this.config);
             log.info("{}: Created Ledger {}.", this.traceObjectId, newLedger.getId());
 
-            // Update node with new ledger.
-            metadata = updateMetadata(metadata, newLedger);
-            LedgerMetadata ledgerMetadata = metadata.getLedger(newLedger.getId());
+            // Update Metadata with new Ledger and persist to ZooKeeper.
+            LogMetadata newMetadata = updateMetadata(oldMetadata, newLedger, true);
+            LedgerMetadata ledgerMetadata = newMetadata.getLedger(newLedger.getId());
             assert ledgerMetadata != null : "cannot find newly added ledger metadata";
             this.writeLedger = new WriteLedger(newLedger, ledgerMetadata);
-            this.logMetadata = metadata;
-            log.info("{}: Initialized.", this.traceObjectId);
+            this.logMetadata = newMetadata;
+            ledgersToDelete = getLedgerIdsToDelete(oldMetadata, newMetadata);
         }
+
+        // Delete the orphaned ledgers from BookKeeper.
```

Comment:
To clarify one thing, in the case that the server crashes right at this point, the orphaned ledgers would remain orphaned forever (unless we manually remove them), is it right? I'm saying this because we update the zk metadata before we remove the ledgers, which makes sense as the other way around gives us a consistency issue.

To be sure that we eventually remove orphaned ledgers, we would need to record the need to remove those ledgers in persistent store until we make sure that they are actually removed.

Reply:
This is a problem we had before and nothing new. Check out #1165 .

Reply:
ok.